File: pt5/1156.swap-for-longest-repeated-character-substring.go
package main

type runGroup struct {
	char   byte
	length int
}

func maxRepOpt1(text string) int {
	total := [26]int{}
	for i := 0; i < len(text); i++ {
		total[text[i]-'a']++
	}

	groups := []runGroup{}
	for i := 0; i < len(text); {
		j := i
		for j < len(text) && text[j] == text[i] {
			j++
		}
		groups = append(groups, runGroup{char: text[i], length: j - i})
		i = j
	}

	best := 0
	for _, group := range groups {
		candidate := group.length
		if total[group.char-'a'] > group.length {
			candidate++
		}
		if candidate > total[group.char-'a'] {
			candidate = total[group.char-'a']
		}
		if candidate > best {
			best = candidate
		}
	}

	for i := 1; i+1 < len(groups); i++ {
		if groups[i].length == 1 && groups[i-1].char == groups[i+1].char {
			char := groups[i-1].char
			combined := groups[i-1].length + groups[i+1].length
			candidate := combined
			if total[char-'a'] > combined {
				candidate++
			}
			if candidate > total[char-'a'] {
				candidate = total[char-'a']
			}
			if candidate > best {
				best = candidate
			}
		}
	}

	return best
}

File: pt5/1156.swap-for-longest-repeated-character-substring.py
from __future__ import annotations

from collections import Counter


class Solution:
    def maxRepOpt1(self, text: str) -> int:
        total = Counter(text)
        groups = []
        i = 0

        while i < len(text):
            j = i
            while j < len(text) and text[j] == text[i]:
                j += 1
            groups.append((text[i], j - i))
            i = j

        best = 0
        for char, length in groups:
            best = max(best, min(length + (total[char] > length), total[char]))

        for i in range(1, len(groups) - 1):
            middle_char, middle_length = groups[i]
            left_char, left_length = groups[i - 1]
            right_char, right_length = groups[i + 1]
            if middle_length == 1 and left_char == right_char:
                combined = left_length + right_length
                best = max(best, min(combined + (total[left_char] > combined), total[left_char]))

        return best

